build-system/server.js

```diff
@@ -98,20 +98,30 @@ const SOURCE_ORIGIN_REGEX = new RegExp('^http://localhost:8000|' +
     '^https?://.+\.herokuapp\.com:8000/');
 
 app.use('/form/html/post', function(req, res) {
-  if (!ORIGIN_REGEX.test(req.headers.origin)) {
-    res.statusCode = 500;
+  if (req.method != 'POST') {
+    res.statusCode = 405;
     res.end(JSON.stringify({
-      message: 'Origin header is invalid.'
+      message: req.method + ' method is not allowed. Use POST.'
     }));
     return;
   }
 
-  if (!SOURCE_ORIGIN_REGEX.test(req.query.__amp_source_origin)) {
-    res.statusCode = 500;
-    res.end(JSON.stringify({
-      message: '__amp_source_origin parameter is invalid.'
-    }));
-    return;
+  if (req.headers['amp-same-origin'] != 'true') {
```

Comment:
In light of the discussion below, we should reverse the checks and process `Origin` header first.